data-prepper-plugins/peer-forwarder/src/main/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProvider.java

```diff
@@ -0,0 +1,210 @@
+
+        private int failedAttemptCount = 0;
+        private volatile ScheduledFuture<?> scheduledDiscovery;
+
+        private AwsCloudMapDynamicEndpointGroup() {
+            eventLoop.execute(this::discoverInstances);
+        }
+
+        private void discoverInstances() {
+            if (isClosing()) {
+                return;
+            }
+
+            DiscoverInstancesRequest discoverInstancesRequest = DiscoverInstancesRequest
+                    .builder()
+                    .namespaceName(namespaceName)
+                    .serviceName(serviceName)
+                    .build();
+
+            LOG.info("Discovering instances.");
+
+            awsServiceDiscovery.discoverInstances(discoverInstancesRequest).whenComplete(
+                    (discoverInstancesResponse, throwable) -> {
+                        if(discoverInstancesResponse != null) {
+                            try {
+                                failedAttemptCount = 0;
+                                updateEndpointsWithDiscoveredInstances(discoverInstancesResponse);
+                            } catch (Throwable ex) {
+                                LOG.warn("Failed to update endpoints.", ex);
+                            }
+                            finally {
+                                scheduledDiscovery = eventLoop.schedule(this::discoverInstances,
+                                        timeToRefreshSeconds, TimeUnit.SECONDS);
+                            }
+                        }
+
+                        if(throwable != null) {
+                            failedAttemptCount++;
+                            final long delayMillis = backoff.nextDelayMillis(failedAttemptCount);
+                            LOG.error("Failed to discover instances for: namespace='{}', serviceName='{}'. Will retry in {} ms.",
+                                    namespaceName, serviceName, delayMillis, throwable);
+
+                            scheduledDiscovery = eventLoop.schedule(this::discoverInstances,
+                                    delayMillis, TimeUnit.MILLISECONDS);
+                        }
+
+                    });
+        }
+
+        private void updateEndpointsWithDiscoveredInstances(DiscoverInstancesResponse discoverInstancesResponse) {
+            List<HttpInstanceSummary> instances = discoverInstancesResponse.instances();
+
+            LOG.info("Discovered {} instances.", instances.size());
+
+            List<Endpoint> endpoints = instances
+                    .stream()
+                    .map(HttpInstanceSummary::attributes)
+                    .map(attributes -> attributes.get("AWS_INSTANCE_IPV4"))
+                    .map(ip -> Endpoint.of(domainName).withIpAddr(ip))
+                    .collect(Collectors.toList());
+
+            setEndpoints(endpoints);
+        }
+
+        @Override
+        protected void doCloseAsync(CompletableFuture<?> future) {
```

Comment:
Will this be invoke when the server is stopped or on SIGKILL signal ?

Reply:
This is large part depends on how the Peer Forwarder uses the list. This is a `Closeable` interface, much like the `DnsAddressEndpointGroup` class. So the caller will need to close it in order for this to run.